boltons/ioutils.py

```diff
@@ -0,0 +1,322 @@
+from io import BytesIO
+from codecs import EncodedFile
+from tempfile import TemporaryFile
+
+# Python2/3 compat
+if sys.version_info[0] == 3:
+    text_type = str
+    binary_type = bytes
+else:
+    text_type = unicode
+    binary_type = str
+
+
+def utf8_encode_if_unicode(string):
+    """
+    Given a string type, return it in bytes, encoding to UTF-8 if the string is
+    unicode.
+    """
+    if isinstance(string, text_type):
+        return string.encode('utf-8')
+    elif isinstance(string, binary_type):
+        return string
+    else:
+        raise TypeError("{} is not a unicode or str object".format(string))
+
+
+class SpooledIOBase(object):
+    """
+    The SpooledTempoaryFile class doesn't support a number of attributes and
+    methods that a StringIO instance does. This brings the api as close to
+    compatible as possible with StringIO so that it may be used as a near
+    drop-in replacement to save memory.
+
+    Another issue with SpooledTemporaryFile is that the spooled file is always
+    a cStringIO rather than a StringIO which causes issues with some of our
+    tools.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, max_size=5000000):
+        self._max_size = max_size
+
+    @abstractmethod
+    def read(self, n=-1):
+        """Read n characters from the buffer"""
+        pass
+
+    @abstractmethod
+    def write(self, s):
+        """Write into the buffer"""
+        pass
+
+    @abstractmethod
+    def readline(self, length=None):
+        """Returns the next available line"""
+        pass
+
+    @abstractmethod
+    def readlines(self, sizehint=0):
+        """Returns a list of all lines from the current position forward"""
+        pass
+
+    @abstractmethod
+    def rollover(self):
+        """Roll file-like-object over into a real temporary file"""
+        pass
+
+    @abstractproperty
+    def buffer(self):
+        """Should return a flo instance"""
+        pass
+
+    @abstractproperty
+    def _rolled(self):
+        """Returns whether the file has been rolled to a real file or not"""
+        pass
+
+    def _get_softspace(self):
+        return self.buffer.softspace
+
+    def _set_softspace(self, val):
+        self.buffer.softspace = val
+
+    softspace = property(_get_softspace, _set_softspace)
+
+    @property
+    def _file(self):
+        return self.buffer
+
+    def tell(self):
+        return self._file.tell()
+
+    def seek(self, pos, mode=0):
+        return self.buffer.seek(pos, mode)
+
+    def close(self):
+        return self.buffer.close()
+
+    def flush(self):
+        return self.buffer.flush()
+
+    def isatty(self):
+        return self.buffer.isatty()
+
+    def next(self):
+        return self.readline()
+
+    @property
+    def closed(self):
+        return self.buffer.closed
+
+    @property
+    def pos(self):
+        return self.tell()
+
+    @property
+    def buf(self):
+        return self.getvalue()
+
+    @property
+    def len(self):
```

Comment:
I agree - implementing __len__ would be trivial too, so there aren't many reasons to leave it out. And yea - I see a lot of StringIO used, particularly in proprietary code. This often leads to horrible memory issues over time, so having a more memory efficient drop-in replacement is quite handy when you need to reduce memory overhead.